packages/contracts-bedrock/scripts/differential-testing.ts

```diff
@@ -0,0 +1,135 @@
+import { BigNumber, utils } from 'ethers'
+import {
+  decodeVersionedNonce,
+  hashCrossDomainMessage,
+  DepositTx,
+  SourceHashDomain,
+  encodeCrossDomainMessage,
+  hashWithdrawal,
+  hashOutputRootProof,
+} from '@eth-optimism/core-utils'
+
+const args = process.argv.slice(2)
+const command = args[0]
+
+switch (command) {
+  case 'decodeVersionedNonce': {
+    const input = BigNumber.from(args[1])
+    const [nonce, version] = decodeVersionedNonce(input)
+
+    const output = utils.defaultAbiCoder.encode(
+      ['uint256', 'uint256'],
+      [nonce.toHexString(), version.toHexString()]
+    )
+    process.stdout.write(output)
+    break
+  }
+  case 'encodeCrossDomainMessage': {
+    const nonce = BigNumber.from(args[1])
+    const sender = args[2]
+    const target = args[3]
+    const value = BigNumber.from(args[4])
+    const gasLimit = BigNumber.from(args[5])
+    const data = args[6]
+
+    const encoding = encodeCrossDomainMessage(
+      nonce,
+      sender,
+      target,
+      value,
+      gasLimit,
+      data
+    )
+
+    const output = utils.defaultAbiCoder.encode(['bytes'], [encoding])
+    process.stdout.write(output)
+    break
+  }
+  case 'hashCrossDomainMessage': {
+    const nonce = BigNumber.from(args[1])
+    const sender = args[2]
+    const target = args[3]
+    const value = BigNumber.from(args[4])
+    const gasLimit = BigNumber.from(args[5])
+    const data = args[6]
+
+    const hash = hashCrossDomainMessage(
+      nonce,
+      sender,
+      target,
+      value,
+      gasLimit,
+      data
+    )
+    const output = utils.defaultAbiCoder.encode(['bytes32'], [hash])
+    process.stdout.write(output)
+    break
+  }
+  case 'hashDepositTransaction': {
+    // The solidity transaction hash computation currently only works with
+    // user deposits. System deposit transaction hashing is not supported.
```

Comment:
Why is that?

Reply:
System deposits compute hash differently based on a different way to compute the source hash

Reply:
OK, I see that in the specs. Missed when that got added.

Reply:
Is it possible we'll need it in solidity anytime soon?

Reply:
> Is it possible we'll need it in solidity anytime soon?

I think we could add it in the future if needed, right now it doesn't solve an end user problem which is why i'm omitting it